### PR TITLE
PLT-4430 improve slow channel switching

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
Remove the getChannels and getMoreChannels requests on channel switch and add a new API for getting the current user's channel members.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-4430

API proposal: https://github.com/mattermost/mattermost-api-reference/pull/12

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Added API documentation (required for all new APIs)
- [x] All new/modified APIs include changes to the drivers
- [x] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

File: api/channel.go [hunk 1]
```diff
@@ -450,6 +451,25 @@ func getChannelCounts(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func getChannelsUnread(c *Context, w http.ResponseWriter, r *http.Request) {
+	if result := <-Srv.Store.Channel().GetChannelsUnread(c.TeamId, c.Session.UserId); result.Err != nil {
+		if result.Err.Id == "store.sql_channel.get_channels.not_found.app_error" {
```
Comment (crspeller) [inline api/channel.go]: Why are we doing this?

File: api/channel_test.go [hunk 2]
```diff
@@ -656,11 +656,11 @@ func TestGetChannel(t *testing.T) {
 	rget := Client.Must(Client.GetChannels(""))
 	data := rget.Data.(*model.ChannelList)
 
-	if data.Channels[0].DisplayName != channel1.DisplayName {
+	if (*data)[0].DisplayName != channel1.DisplayName {
```
Comment (crspeller) [inline api/channel_test.go]: Maybe rename some of these to `channels` or something to make it more clear.

File: api/channel_test.go [hunk 3]
```diff
@@ -1261,13 +1314,13 @@ func TestUpdateNotifyProps(t *testing.T) {
...
 	}
 
-	rget := Client.Must(Client.GetChannels(""))
-	rdata := rget.Data.(*model.ChannelList)
-	if len(rdata.Members) == 0 || rdata.Members[channel1.Id].NotifyProps["desktop"] != data["desktop"] {
-		t.Fatal("NotifyProps[\"desktop\"] did not update properly")
-	} else if rdata.Members[channel1.Id].LastUpdateAt <= timeBeforeUpdate {
-		t.Fatal("LastUpdateAt did not update")
-	}
+	//rget := Client.Must(Client.GetChannels(""))
```
Comment (crspeller) [inline api/channel_test.go]: Why not delete?

File: api/channel.go [hunk 4]
```diff
@@ -21,6 +21,7 @@ func InitChannel() {
 	BaseRoutes.Channels.Handle("/", ApiUserRequired(getChannels)).Methods("GET")
 	BaseRoutes.Channels.Handle("/more", ApiUserRequired(getMoreChannels)).Methods("GET")
 	BaseRoutes.Channels.Handle("/counts", ApiUserRequired(getChannelCounts)).Methods("GET")
+	BaseRoutes.Channels.Handle("/unread", ApiUserRequired(getChannelsUnread)).Methods("GET")
```
Comment (crspeller) [inline api/channel.go]: Why call this `/unread` and `getChannelsUnread`? It's more like `getChannelMembers`.

File: webapp/actions/global_actions.jsx [hunk 5]
```diff
@@ -42,8 +42,7 @@ export function emitChannelClickEvent(channel) {
         );
     }
     function switchToChannel(chan) {
-        AsyncClient.getChannels(true);
-        AsyncClient.getMoreChannels(true);
+        AsyncClient.getChannelsUnread(true);
```
Comment (crspeller) [inline webapp/actions/global_actions.jsx]: Is this really something that needs to happen every time we switch channels? Shouldn't this update though the websocket?

File: api/channel.go (hunk 4 above)
Comment (jwilander) [inline api/channel.go]: Agreed, I'll flip it to make it match the team members version of this

**Comment by jwilander** [comment]
@crspeller please re-review at your leisure

**Pull request merged**